### ISSUE TITLE
[ProjectTracking]: Eth wallet support

### ISSUE THREAD

**Issue by walnut-the-cat** [near/near-one-project-tracking#7]
# Goals

## What is the project trying to achieve?
We want the NEAR Protocol's ecosystem to be closer with Ethereum by integrating Web3 wallet support. To accomplish that, some changes on the protocol level are needed with an emphasis on user experience continuity.

## Why should the Near One team work on this project?
Currently, the Ethereum ecosystem is a leading force in the smart contract blockchain space, boasting a large user base and extensive installations of Ethereum-compatible tooling and wallets. However, a significant challenge arises due to the incompatibility of these tools and wallets with NEAR Protocol. This incompatibility necessitates a complete onboarding process for users to interact with NEAR contracts and accounts, leading to confusion, decreased adoption, and the marginalization of NEAR Protocol.

## How does it improve the NEAR protocol?
Implementing Web3 wallet support in NEAR Protocol, with an emphasis on user experience continuity, would significantly benefit the entire NEAR Ecosystem. It enhances NEAR's interoperability with Ethereum, significantly improving accessibility and user experience for a broader audience in the blockchain community.

## What needs to be accomplished?
Please refer to [NEP-518](https://github.com/near/NEPs/issues/518).

Following the design, the main change on the protocol level would be to have ETH-style addresses managed by a new Wallet Contract integrated into the protocol.
For seamless onboarding of EVM users, the contract will be free of charge.

# Links to external documentations and discussions
* [Main nearcore tracking issue](https://github.com/near/nearcore/issues/10018)
* [NEP-518](https://github.com/near/NEPs/issues/518)
  * [Design Review Deck](https://docs.google.com/presentation/d/1hly3r1EQQWMV6MEk2WlUUfOIMgyMCeYy2LW9BVoVX4U/edit?usp=sharing)
  * [Design Review Call Recording](https://drive.google.com/file/d/1AXN8xZE5KfsPL84LewAtRolzyLIBQIep/view?usp=sharing)
* [Zulip discussion](https://near.zulipchat.com/#narrow/stream/295558-pagoda.2Fcore/topic/Ethereum.20transaction.20.26.20wallet.20support)

# Estimated effort
* [`T-core`, `T-contract-runtime`] Estimated time by which teams in Near One will need to contribute to this project (engineering time) is the end of 2023.
* [`T-core`, `T-Aurora`] Estimated time to finish the project (release time) is the end of Q1 2024, and will involve auditing the `Wallet Contract`.
* [`T-Aurora`, `T-integrations`] Estimated time to fully deliver the project (out of Near One scope) is past Q1 2024, as it involves a lot of integration, instructing validators, users, etc.

# Assumptions
- Some RPC nodes will have to host `RPC-Translator` (to provide Ethereum methods implementations using the NEAR RPC as a data source) and a `Transaction Relayer` that embeds Ethereum-compatible transactions into a NEAR transaction and forwards it to the user’s Wallet Contract.

# Pre-requisites
- The creation of Ethereum addresses on NEAR has been restricted: https://github.com/near/NEPs/pull/492
- This project started after the initial design has been proposed: https://github.com/near/NEPs/issues/498
- Since the start of this project the current design has been proposed: https://github.com/near/NEPs/issues/518

# Out of scope
Just the protocol level changes for ETH Wallet support are part of this project. Things that are out of scope of this repository (e.g. `RPC-Translator`) are out of scope of this project and are owned by the Aurora team.

**Comment by walnut-the-cat** [comment]
Implementation for Eth-implicit addresses support is complete according to the new design, but we want to treat Wallet Contract specially and not deploy it directly (a PR for that is in progress). The implementation of Wallet Contract is on the Aurora side, they will likely make it till the early January, after some paper work is finished. NEP is under draft, after some discussions it looks it won’t change.

**Comment by staffik** [comment]
* [The PR to treat the Wallet Contract specially](https://github.com/near/nearcore/pull/10269) is under review.
* We might want to add support for the Mirror tool and some Nayduck tests.
* There is nothing more we can do before the real Wallet Contract is ready.

**Comment by staffik** [comment]
December 19th status update
* In December we did [Wallet Contract placeholder](https://github.com/near/nearcore/pull/10269). That finishes protocol changes for this project. As a follow up PR, we want to make small operational changes:
  * Include `wallet-contract` crate to the workspace (blocked by `near-sdk` having vulnerable dependencies).
  * Use [Docker container](https://hub.docker.com/r/nearprotocol/contract-builder) instead of `cargo build` command to generate WASM file.
* The above should be implemented this week, but waiting for new `near-sdk` version to be released.

**Comment by staffik** [comment]
January 11th status update
* All the protocol changes are merged, needs some tooling related follow-up PRs:
  * generating WASM file using NEAR docker container
  * including the wallet contract into the workspace, so that the linting works there
  * making sure the mirror tool works well
* Waiting for the real Wallet Contract implementation but nothing happened on the Aurora side yet.
* Project is currently on hold on the Protocol side, will update in 1 month if anything changes.

**Comment by staffik** [comment]
2024 February 5th status update
* nothing happened, waiting for the Aurora part, will have a sync this month
